### PR TITLE
Improve BYTE_ARRAY dictionaries

### PR DESCRIPTION
**Pull request by achille-roussel**
This PR contributes to #226 by optimizing the `byteArrayDictionary` type.

I made multiple attempts to improve throughput of probing operations for string values (reflected in this branch history) but without being able to significantly increase throughput in the way that I did for probing fixed-length values (e.g. int32). The increased complexity did not justify the returns, so we will stick with `map[string]int32` in this case. 

Overall, measurable but minor improvements, bug fixes, and simplification of the code.

```
name                                    old time/op  new time/op  delta
Dictionary/Insert/BYTE_ARRAY/N=100      1.40µs ± 0%  1.36µs ± 1%  -2.88%  (p=0.000 n=8+9)
Dictionary/Insert/BYTE_ARRAY/N=1000     13.8µs ± 1%  13.6µs ± 1%  -1.73%  (p=0.000 n=10+9)
Dictionary/Insert/BYTE_ARRAY/N=10000     414µs ± 2%   403µs ± 0%  -2.72%  (p=0.000 n=10+10)
Dictionary/Insert/BYTE_ARRAY/N=100000   5.58ms ± 4%  5.57ms ± 2%    ~     (p=0.579 n=10+10)
Dictionary/Insert/BYTE_ARRAY/N=1000000   105ms ± 2%   105ms ± 2%    ~     (p=0.436 n=10+10)

name                                    old value/s  new value/s  delta
Dictionary/Insert/BYTE_ARRAY/N=100       71.3M ± 1%   73.5M ± 1%  +3.04%  (p=0.000 n=9+9)
Dictionary/Insert/BYTE_ARRAY/N=1000      72.3M ± 1%   73.6M ± 1%  +1.77%  (p=0.000 n=10+9)
Dictionary/Insert/BYTE_ARRAY/N=10000     24.1M ± 2%   24.8M ± 0%  +2.79%  (p=0.000 n=10+10)
Dictionary/Insert/BYTE_ARRAY/N=100000    17.9M ± 4%   18.0M ± 2%    ~     (p=0.579 n=10+10)
Dictionary/Insert/BYTE_ARRAY/N=1000000   9.56M ± 2%   9.51M ± 2%    ~     (p=0.436 n=10+10)
```

### CODE REVIEW

File: dictionary.go [hunk 1]
```diff
@@ -709,36 +711,46 @@ func (d *byteArrayDictionary) Insert(indexes []int32, values []Value) {
...
-func (d *byteArrayDictionary) append(value string) string {
-	offset := len(d.values)
-	d.values = plain.AppendByteArrayString(d.values, value)
-	d.offsets = append(d.offsets, uint32(offset))
-	d.numValues++
-	return string(d.values[offset+plain.ByteArrayLengthSize : len(d.values)])
+func cloneString(s string) string {
+	b := new(strings.Builder)
+	b.WriteString(s)
+	return b.String()
```
Comment (kevinburkesegment) [inline dictionary.go:753]: This can't be the best way to do this, can it...?

Comment (kevinburkesegment) [reply dictionary.go:753]: sigh... 

```
// String returns the accumulated string.
func (b *Builder) String() string {
	return *(*string)(unsafe.Pointer(&b.buf))
}

File: sparse/gather_amd64.go [hunk 2]
```diff
@@ -81,3 +85,11 @@ func gather64AVX2(dst []uint64, src Uint64Array)
 
 //go:noescape
 func gather128(dst [][16]byte, src Uint128Array) int
+
+// On AMD64 the `string` type is made of a 64 bit pointer and a 64 bit integer,
+// for a total of 128 bit which means that we can use the same algorithm as
+// gather128 to lookup string values, the memory layout of the arguments being
```
Comment (kevinburkesegment) [inline sparse/gather_amd64.go:91]:
```suggestion
// gather128 to lookup string values, since the memory layout of the arguments is
```

**Review by kevinburkesegment — APPROVED** [review]

File: dictionary.go (hunk 1 above)
Comment (achille-roussel) [reply dictionary.go:753]: Go 1.18 has [`strings.Clone](https://cs.opensource.google/go/go/+/refs/tags/go1.18.4:src/strings/clone.go;l=21-28) but we have to maintain compatibility with Go 1.17 for now.

**Comment by achille-roussel** [comment]
Thanks for the review!

**Pull request merged**